### ISSUE TITLE
Intellisense warning: Not all code paths return a value. I think it shouldnot show the warning in my code snippet.

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#47433]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/874987/intellisense-warning-not-all-code-paths-return-a-v.html)._

---
I have a code snippet:

enum E
{
	A, B, C
}

class Test
{
	int Method(E e)
	{
		switch (e)
		{
			case E.A: return 0;
			case E.B: return 1;
			case E.C: return 2;
		}
	}
}

There will be an error at "Method" line: "Not all code paths return a value".
But i think the compiler/the IDE should know exactly that the Method WILL ALWAYS RETURN A VALUE.
And therefore there will have no error at all.
Sorry if i was wrong. Thank you for reading

---
#### Feedback Bot on 1/6/2020, 00:47 AM: 

Thank you for taking the time to provide your suggestion.  We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.

**Comment by PathogenDavid** [comment]
This method will not always return a value because you can coerce arbitrary integers to be enums. For example:

```csharp
Method((E)99);
```

The solution you can use today is to do something sensible when an invalid enum value is provided:

```csharp
int Method(E e)
{
    switch (e)
    {
        case E.A: return 0;
        case E.B: return 1;
        case E.C: return 2;
        default: throw new ArgumentException("The specified enum is invalid.", nameof(e));
    }
}
```

There is a language feature proposal tentatively scheduled for C#10 to add closed enums to the language, which act more like you expect: https://github.com/dotnet/csharplang/issues/3179

**Comment by CyrusNajmabadi** [comment]
Closing out as by design as per the language.  Feel free to track/vote on the linked language issue on closed-enum sets if that interests you.
